source/math/FloatVectorList.ooc

```diff
@@ -500,4 +310,194 @@ FloatVectorList: class extends VectorList<Float> {
+	}
+	backwardGaussianKernel: static func ~defaultSigma (size: Int) -> This {
+		This backwardGaussianKernel(size, size as Float / 3.0f)
+	}
+	backwardGaussianKernel: static func ~full (size: Int, sigma: Float) -> This {
+		result := This new(size)
+		forwardGaussian := This forwardGaussianKernel(size, sigma)
+		for (i in 0 .. size)
+			result add(forwardGaussian[size - i - 1])
+		forwardGaussian free()
+		result
+	}
+	_swap: static func (array: Float*, i, j: Int) {
+		t := array[i]
+		array[i] = array[j]
+		array[j] = t
+	}
+	/* partition the array so that all elements less than array[pivot] are moved before this element
+		and all elements greater than array[pivot] are after it
+		end should be the last valid array index in the range
+	*/
+	_partition: static func (array: Float*, start, end, pivot: Int) -> Int {
+		pivotValue := array[pivot]
+		This _swap(array, pivot, end)
+		result := start
+		for (i in start .. end)
+			if (array[i] < pivotValue) {
+				This _swap(array, result, i)
+				++result
+			}
+		This _swap(array, result, end)
+		result
+	}
+	/*
+	pivot selection strategy which should result in best performance
+	for quicksort and nthElement algorithms on partially sorted data
+		end should be the last valid array index in the range
+	*/
+	_medianOfThree: static func (array: Float*, start, end: Int) -> Int {
+		mid := (start + end) / 2
+		if (array[start] > array[mid])
+			This _swap(array, mid, start)
+		if (array[mid] > array[end])
+			This _swap(array, mid, end)
+		if (array[start] > array[end])
+			This _swap(array, start, end)
+		mid
+	}
+
+	/* return n-th smallest element in the array[start:end] range
+		end should be the last valid array index in the range
+	*/
+	_nthElement: static func (array: Float*, start, end, n: Int) -> Float {
+		if (start == end)
+			array[start]
+		else {
+			pivot := This _partition(array, start, end, This _medianOfThree(array, start, end))
+			if (pivot == n)
+				array[n]
+			else if (n < pivot)
+				_nthElement(array, start, pivot - 1, n)
+			else
+				_nthElement(array, pivot + 1, end, n)
+		}
+	}
+
+	/* sort the input array in the range [start,end]
+		end should be the last valid array index in the range
+	*/
+	_quicksort: static func (array: Float*, start, end: Int) {
+		if (end == start + 1 && array[start] > array[end])
+			This _swap(array, start, end)
+		else if (start < end) {
+			pivot := This _partition(array, start, end, This _medianOfThree(array, start, end))
+			if (pivot > start)
+				This _quicksort(array, start, pivot - 1)
+			if (pivot < end)
+				This _quicksort(array, pivot + 1, end)
+		}
+	}
+	_generateCubicSplineCoefficients: static func (size: Int) -> FloatMatrix {
+		result := FloatMatrix new(size, size) take()
+		result[0, 0] = 2.f
+		result[0, 1] = 1.f
+		for (index in 1 .. size - 1) {
+			result[index, index - 1] = 1.f
+			result[index, index] = 4.f
+			result[index, index + 1] = 1.f
+		}
+		result[size - 1, size - 2] = 1.f
+		result[size - 1, size - 1] = 2.f
+		result
+	}
```

Comment:
There's some cleanup that can be done in the above functions. Don't think `quicksort` should be static and the comments aren't very useful. `_nthElement` should be `_nthSmallestElement`. Writing this here so you will note it down and one of us will fix this after this PR.